### ISSUE TITLE
在 MacOS 上，http自定义 API 仅支持 localhost，使用非本机 ip 或 url 均无法访问[Bug] 

### ISSUE THREAD

**Issue by wjfu99** [ChatGPTNextWeb/ChatGPT-Next-Web#4095]
### Bug Description

自定义接口使用本地部署的 API可以正常使用，使用非本机部署的API 显示 failed to load 或 failed to fetch。部署的 API 应该没有问题，在本机通过端口转发将外部 ip 对应端口映射至本地也可正常使用。

### Steps to Reproduce

1.使用本地部署 api：
<img width="1847" alt="image" src="https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/57850011/4c98eea2-9405-4e30-b24c-8dd5e29945f6">

2.使用非本机部署 api：
<img width="939" alt="image" src="https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/57850011/04d97056-d012-44d7-b68f-07f508c0bef0">

### Expected Behavior

使用 http+外部 ip 的 API 也可以正常使用。

### Deployment Method

- [X] Docker
- [ ] Vercel
- [ ] Server

### Desktop OS

Macos

**Comment by fred-bf** [comment]
from you screenshot it seems to the cors issue with your openai api provider?

**Comment by wjfu99** [comment]
> from you screenshot it seems to the cors issue with your openai api provider?

but if i forward the api provider ip&port to local, then the nextweb works well, it's very strange. By the way, the api provider is deployed in a server on the same subnet, and i currently use vscode for prot forwarding.😂

**Comment by wjfu99** [comment]
sorry, close the issue by accident.

**Comment by wjfu99** [comment]
> from you screenshot it seems to the cors issue with your openai api provider?

Besides, I can access the API service well via curl command:
```
curl --location 'http://ip:port/v1/chat/completions' \
--header 'Content-Type: application/json' \
--header 'Authorization: xxxxx' \
--data '{
  "model": "gpt-4",
  "messages": [ {"role": "user", "content": "hello"}]
}'
```
the api service also works with a congeneric app [ChatX](https://apps.apple.com/us/app/chatx-ai-chat-client/id6446304087), but in NextChat, it still not work.

**Comment by wjfu99** [comment]
I have used Burp Suite to capture http package, and the result demonstrate that when using a API deployed in localhost, the request packages are sent correctly:

```
OPTIONS /v1/chat/completions HTTP/1.1
Host: localhost:1999
Accept: */*
Access-Control-Request-Method: POST
Access-Control-Request-Headers: content-type,x-requested-with
Origin: https://app.nextchat.dev
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.6167.160 Safari/537.36
Sec-Fetch-Mode: cors
Sec-Fetch-Site: cross-site
Sec-Fetch-Dest: empty
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Connection: close
```
```
POST /v1/chat/completions HTTP/1.1
Host: localhost:1999
Content-Length: 773
sec-ch-ua: "Chromium";v="121", "Not A(Brand";v="99"
Accept: application/json, text/event-stream
Content-Type: application/json
x-requested-with: XMLHttpRequest
sec-ch-ua-mobile: ?0
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.6167.160 Safari/537.36
sec-ch-ua-platform: "macOS"
Origin: https://app.nextchat.dev
Sec-Fetch-Site: cross-site
Sec-Fetch-Mode: cors
Sec-Fetch-Dest: empty
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Connection: close

{"messages":[{"role":"system","content":"\nYou are ChatGPT, a large language model trained by OpenAI.\nKnowledge cutoff: 2021-09\nCurrent model: gpt-4\nCurrent time: 2/22/2024, 9:31:13 AM\nLatex inline: $x^2$ \nLatex block: $$e=mc^2$$\n\n"},{"role":"user","content":"hello"},{"role":"assistant","content":"Hello! How can I assist you today?"},{"role":"user","content":"0.0.0.0 是监听本机所有 ip 吗"},{"role":"assistant","content":"\n\nUnauthorized access, please enter access code in [auth](/#/auth) page, or enter your OpenAI API Key.\n\n```json\n{\n  \"error\": \"Unauthorized\"\n}\n```"},{"role":"user","content":"docker -e 可以添加多个环境变量吗"}],"stream":true,"model":"gpt-4","temperature":0.5,"presence_penalty":0,"frequency_penalty":0,"top_p":1}
```

And if I set the endpoint API to another subnet host (Like http://222.21.136.207:1999), then no http package is captured, and the NextChat just raise an error:
![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/57850011/ebbe1a58-b5ab-4b2d-8266-5aca98604f8f)

Besides, set the API to localhost, then capture and edit the http package to the subnet host, the NextChat can works correctly.

**Comment by SubtleSpark** [comment]
你是不是使用了 NAT 进行了端口映射，这可能会有 NAT 环回的问题。

**Comment by wjfu99** [comment]
> 你是不是使用了 NAT 进行了端口映射，这可能会有 NAT 环回的问题。
我应该没有主动配置 NAT 端口映射，而且 NAT 环回会导致 NextChat这类 app不能 send url request 吗？主要用 curl又是好的，就很奇怪。

**Comment by WAGMY** [comment]
> Same problem, Macos + Docker, Set api to localhost works fine, but it won't work if set to my own server ip, but I can access the api service well via python script.

Supplement: If using windows nextchat client, it work fine if set to my own server ip

**Comment by wjfu99** [comment]
> > Same problem, Macos + Docker, Set api to localhost works fine, but it won't work if set to my own server ip, but I can access the api service well via python script.
> 
> Supplement: If using windows nextchat client, it work fine if set to my own server ip

Very useful information, i'm a mac user too, and the nextchat web also didn't work on ios device. i will test it on windows later.

**Comment by daobee** [comment]
Same problem. 
I hosted a OpenAI-Like API on server and it works fine with curl(on macos)/Next-Web(on windows client),but raised a error using Next-Web(on macos client)

curl(on macos):
![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/21331325/8576ece5-683c-494b-a3a4-2117c0cd793c)
Next-Web(on windows client):
Just works fine.
Next-Web(on macos client):
![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/21331325/46df2373-ca0c-4cb8-aa77-5573302d3bf4)

**Comment by wjfu99** [comment]
@fred-bf I believe the bug tag should be added back, since another issue #2810 reveals this bug too. This bug seems to exist for a period.

**Comment by wjfu99** [comment]
> I don't have any issues with this; it has been running on Kubernetes (k8s) without problems

Are you using an Mac client? It seems that the issue only exist on Mac.

**Comment by H0llyW00dzZ** [comment]
> > I don't have any issues with this; it has been running on Kubernetes (k8s) without problems
> 
> Are you using an Mac client? It seems that the issue only exist on Mac.

right now I am using windows client and using ip pods k8s

**Comment by likeman66** [comment]
kind reminder, ur api domain exposed in ur first image

**Comment by fred-bf** [comment]
I just tried build a development macOS client,  it could make request to custom ip. But when switching to a online version, I could reproduce this issue. It might be released to tauri production build configuration?

**Comment by H0llyW00dzZ** [comment]
> I just tried build a development macOS client, it could make request to custom ip. But when switching to a online version, I could reproduce this issue. It might be released to tauri production build configuration?

no,the problem its not only desktop

![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/17626300/0138d1de-f9a6-4619-9edb-e52788c0f2f8)

**Comment by wjfu99** [comment]
> I just tried build a development macOS client, it could make request to custom ip. But when switching to a online version, I could reproduce this issue. It might be released to tauri production build configuration?

Supplement: the same problem also occurs on the online web version ([app.nextchat.dev](https://app.nextchat.dev/)) and the self-deployed vercel version. Hope this info help.

**Comment by wjfu99** [comment]
@fred-bf Hi, you are absolutely right! i check the console log, and find the mixed content error:
![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/57850011/9e50bef9-61e0-4a55-bf3c-041c9001a8a9)
So, this issue on web version can be solved by allowing the insecure content in website permission:
<img width="832" alt="image" src="https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/57850011/5e87babd-2694-44ca-9297-96ebd44ef0ff">
By the way, do you have any ideas on how to solve this problem in the local client version?🤣


> @wjfu99 For web version, plz check whether its blocked by mixed https content ![CleanShot 2024-02-26 at 12 39 17@2x](https://private-user-images.githubusercontent.com/157469842/307665848-b1fff87e-fd4c-4558-bb74-26db646b3da8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5MjY5MDEsIm5iZiI6MTcwODkyNjYwMSwicGF0aCI6Ii8xNTc0Njk4NDIvMzA3NjY1ODQ4LWIxZmZmODdlLWZkNGMtNDU1OC1iYjc0LTI2ZGI2NDZiM2RhOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyNlQwNTUwMDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNjY1NzM3NGExMWZiYjVkY2NkYTQ1YmRiNjUyZWExNTAwMmFkYWE2ZDIyOTY2ODFkOGJjOGEzMzQ1NWE1YjkzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Bpjf4n0E0vNO1AQoYnGNBB6e02yyEn6EvozRNOfyclc)

**Comment by fred-bf** [comment]
close for now, please feel free to report if you encounter any additional issue

**Comment by hhhaiai** [comment]
how fix
![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/5738285/0e3a05d4-da42-4056-85df-6fe609d76cef)

**Comment by hhhaiai** [comment]
how can i do, how fix
> > 机器人检测到问题主体的语言不是英语，请自动翻译。
> 
> 如何修复 ![image](https://private-user-images.githubusercontent.com/5738285/327954880-0e3a05d4-da42-4056-85df-6fe609d76cef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ4MzMxMjAsIm5iZiI6MTcxNDgzMjgyMCwicGF0aCI6Ii81NzM4Mjg1LzMyNzk1NDg4MC0wZTNhMDVkNC1kYTQyLTQwNTYtODVkZi02ZmU2MDlkNzZjZWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDRUMTQyNzAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWUxYjNiNmUxNTRiMTJhY2Q5MDY2NWYyZGMwY2Q4MDE5ZGFjMzFiYmU4ZWJmNjJhZjFiYTYzOGVmNmZjYjBlZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Ma7l5oByAFB58FYLkdrmTJqoY3wcsODa_fFWS1s4bG0)

**Comment by wjfu99** [comment]
@hhhaiai you can solve this by allowing the insecure content in website permission (chrome and edge should both support that): [refer to this reply](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/issues/4095#issuecomment-1963380038)

**Comment by chengs** [comment]
为什么要发送OPTIONS给API？直接发POST不行么
